R/sampler-hmc.R

```diff
@@ -0,0 +1,220 @@
+    compute_gradient <- function(x) {
+      internal$transform$deriv(x) * model$gradient(x)
+    }
+
+    theta <- internal$transform$model2rn(state$pars)
+    pars_next <- state$pars
+    theta_next <- theta
+
+    if (debug) {
+      history <- matrix(NA_real_, n_integration_steps + 1, length(theta))
+      history[1, ] <- pars_next
+    }
+
+    v <- internal$sample_momentum(rng)
+
+    ## Make a half step for momentum at the beginning
+    v_next <- v + epsilon * compute_gradient(pars_next) / 2
+
+    for (i in seq_len(n_integration_steps)) {
+      ## Make a full step for the position
+      theta_next <- theta_next + epsilon * v_next
+      pars_next <- internal$transform$rn2model(theta_next)
+      ## Make a full step for the momentum, except at end of trajectory
+      if (i != n_integration_steps) {
+        v_next <- v_next + epsilon * compute_gradient(pars_next)
+      }
+      if (debug) {
+        history[i + 1, ] <- pars_next
+      }
+    }
+
+    ## Make a half step for momentum at the end.
+    v_next <- v_next + epsilon * compute_gradient(pars_next) / 2
+
+    ## Some treatments would negate momentum at end of trajectory to
+    ## make the proposal symmetric by setting v_next <- -v_next but
+    ## this is not actually needed in practice.
+
+    ## Potential energy at the end of the trajectory
+    density_next <- model$density(pars_next)
+
+    ## Kinetic energies at the start (energy) and end (energy_next) of
+    ## the trajectories
+    energy <- sum(v^2) / 2
+    energy_next <- sum(v_next^2) / 2
+
+    ## Accept or reject the state at end of trajectory, returning
+    ## either the position at the end of the trajectory or the initial
+    ## position
+    u <- rng$random_real(1)
+    accept <- u < exp(density_next - state$density + energy - energy_next)
+    if (accept) {
+      state$density <- density_next
+      state$pars <- pars_next
+    }
+
+    if (debug) {
+      internal$history <-
+        c(internal$history, list(list(pars = history, accept = accept)))
+    }
+
+    state
+  }
+
+  finalise <- function(state, model, rng) {
+    if (debug) {
+      pars <- lapply(internal$history, "[[", "pars")
+      pars <- array(unlist(pars),
+                    c(dim(pars[[1]]), length(pars)),
+                    list(NULL, model$parameters, NULL))
+      accept <- vlapply(internal$history, "[[", "accept")
+      list(debug = list(pars = pars, accept = accept))
+    } else {
+      NULL
+    }
+  }
+
+  mcstate_sampler("Hamiltonian Monte Carlo",
+                  initialise,
+                  step,
+                  finalise)
+}
+
+
+hmc_transform <- function(domain) {
+  lower <- domain[, 1]
+  upper <- domain[, 2]
+
+  is_bounded <- is.finite(lower) & is.finite(upper)
+  is_semi_infinite <- is.finite(lower) & upper == Inf
+  is_infinite <- lower == -Inf & upper == Inf
+  unknown <- !(is_bounded | is_semi_infinite | is_infinite)
+  if (any(unknown)) {
+    ## This is not very likely to be thrown
```

Comment:
So this would be... 
lower == upper == (-Inf)
lower == upper == (Inf)
or
lower = -Inf and upper is finite   ? (Or is this semi-infinite?)

Reply:
`(-Inf, 0]` would error here, yeah.  But I don't know of any distributions like that!  We can support that if needed fairly simply but no point complicating things until we need to